### PR TITLE
Create a generic type for the data param

### PR DESCRIPTION
**Pull request by christiancp100**
### What does it do?

Describe the technical changes you did.

### Why is it needed?

Describe the issue you are solving.

### How to test it?

Provide information about the environment and the path to verify the behaviour.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

**Comment by Convly** [comment]
Hey, could you update your branch with the `features` one, please? It's a bit hard to see the exact changes that you've done right now

File: packages/core/strapi/lib/services/entity-service/types/params/index.d.ts [hunk 1]
```diff
@@ -4,11 +4,12 @@ import type * as Sort from './sort';
...
+import type * as AttributeUtils from './attributes';
 
 export type For<TSchemaUID extends Common.UID.Schema> = {
   sort?: Sort.Any<TSchemaUID>;
   fields?: Fields.Any<TSchemaUID>;
   filters?: Filters.Any<TSchemaUID>;
 } & Pagination.Any;
 
-export type { Sort, Pagination, Fields, Filters };
+export type { Sort, Pagination, Fields, Filters, AttributeUtils };
```
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/types/params/index.d.ts:15]: I don't think we need to re-export those utils outside of the params for now

File: packages/core/strapi/lib/services/entity-service/index.d.ts [hunk 2]
```diff
@@ -19,8 +21,12 @@ type PaginationInfo = {
   total: number;
 };
 
-type Params<T> = {
-  fields?: (keyof T)[];
+type DataParam<TSchemaUID extends Common.UID.Schema> = {
+  [Key in Attribute.GetKeys<TSchemaUID>]?: AttributeUtils.GetValue<Attribute.Get<TSchemaUID, Key>>;
+};
+
+type Params<TSchemaUID extends Common.UID.Schema> = {
```
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/index.d.ts:28]: This isn't the right type to update :)

packages/core/strapi/lib/services/entity-service/types/params/index.d.ts

File: packages/core/strapi/lib/services/entity-service/index.d.ts [hunk 3]
```diff
@@ -19,8 +21,12 @@ type PaginationInfo = {
   total: number;
 };
 
-type Params<T> = {
-  fields?: (keyof T)[];
+type DataParam<TSchemaUID extends Common.UID.Schema> = {
+  [Key in Attribute.GetKeys<TSchemaUID>]?: AttributeUtils.GetValue<Attribute.Get<TSchemaUID, Key>>;
```
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/index.d.ts:25]: What about required fields?
Shouldn't they be required in here too?

Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/index.d.ts:25]: I'm wondering if we could find a better name for this, maybe something like Payload? Or Input?

Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/index.d.ts:25]:
```suggestion
  [TKey in Attribute.GetKeys<TSchemaUID>]?: AttributeUtils.GetValue<Attribute.Get<TSchemaUID, TKey>>;
```

File: packages/core/strapi/lib/services/entity-service/index.d.ts [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import type { Database, ID } from '@strapi/database';
-import { Strapi } from '../../';
+import { Attribute, Common, Strapi } from '../../';
```
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/index.d.ts:2]:
```suggestion
import { Attribute, Common, Strapi } from '@strapi/strapi';
```

File: packages/core/strapi/lib/services/entity-service/index.d.ts [hunk 5]
```diff
@@ -73,11 +79,14 @@ export interface EntityService {
   ): Promise<T>;
 
   count<K extends keyof AllTypes, T extends AllTypes[K]>(uid: K, params: Params<T>): Promise<any>;
-  create<K extends keyof AllTypes, T extends AllTypes[K]>(uid: K, params: Params<T>): Promise<any>;
-  update<K extends keyof AllTypes, T extends AllTypes[K]>(
-    uid: K,
+  create<TSchemaUID extends Common.UID.ContentType>(
```
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/index.d.ts:82]: Let's wait before updating those types (they're part of other tasks)

(even though I guess it was for testing purpose, let's try to keep the changes atomic for now)

File: packages/core/strapi/lib/services/entity-service/types/params/attributes.d.ts [hunk 6]
```diff
@@ -82,6 +82,13 @@ export type GetValue<TAttribute extends Attribute.Attribute> = Utils.Expression.
         Utils.Expression.Extends<TAttribute, Attribute.Timestamp | Attribute.DateTime>,
         DateTimeValue
       ],
+      [
+        Utils.Expression.Extends<
+          TAttribute,
+          Attribute.Relation<infer TOrigin, infer TRelationKind, infer TTarget>
+        >,
+        Attribute.RelationValue<TRelationKind, TTarget>
```
Comment (Convly) [inline packages/core/strapi/lib/services/entity-service/types/params/attributes.d.ts:90]: Does this work in the end? Or is it still wip?

File: packages/core/strapi/lib/services/entity-service/index.d.ts (hunk 2 above)
Comment (christiancp100) [reply packages/core/strapi/lib/services/entity-service/index.d.ts:28]:
```
export type For<TSchemaUID extends Common.UID.Schema> = {
  sort?: Sort.Any<TSchemaUID>;
  fields?: Fields.Any<TSchemaUID>;
  filters?: Filters.Any<TSchemaUID>;
  data?: Data.Any<TSchemaUID>;
} & Pagination.Any;
```

File: packages/core/strapi/lib/services/entity-service/index.d.ts (hunk 5 above)
Comment (christiancp100) [reply packages/core/strapi/lib/services/entity-service/index.d.ts:82]: But the task in Jira says: `When accessing the create or update methods then I want types the data parameter to be accessible`, if we don't update this, they won't be accessible

**Pull request closed without merging**